prov/util/src/util_ns.c

```diff
@@ -540,57 +517,53 @@ void *ofi_ns_resolve_name(struct util_ns *ns, const char *server_hostname,
  * Name server API: server side
  */
 
-void ofi_ns_start_server(struct util_ns *ns)
+int ofi_ns_start_server(struct util_ns *ns)
 {
 	int ret;
-	SOCKET sockfd;
-	int sleep_usec = 1000;
-	char *server_hostname = (ns->hostname ? ns->hostname :
-				 OFI_NS_DEFAULT_HOSTNAME);
 
-	if ((!ns->is_initialized) || (ofi_atomic_inc32(&ns->ref) > 1))
-		return;
+	assert(ns->is_initialized);
+	if (ofi_atomic_inc32(&ns->ref) > 1)
+		return 0;
 
-	ret = pthread_create(&ns->thread, NULL,
-			     util_ns_name_server_func, (void *)ns);
-	if (ret) {
-		/*
-		 * use the main thread's ID as invalid
-		 * value for the new thread
-		 */
-		ns->thread = pthread_self();
+	ns->map = rbtNew(ns->service_cmp);
+	if (!ns->map) {
+		ret = -FI_ENOMEM;
+		goto err1;
 	}
 
-	/*
-	 * Wait for the local name server to come up. It could be the thread
-	 * created above, or the thread created by another process on the same
-	 * node. The total wait time is about (1+2+4+...+8192)ms = 16 seconds.
-	 */
-	while (sleep_usec < 10000) {
-		sockfd = util_ns_connect_server(ns, server_hostname);
-		if (sockfd != INVALID_SOCKET) {
-			ofi_close_socket(sockfd);
-			return;
-		}
-		usleep(sleep_usec);
-		sleep_usec *= 2;
-	}
```

Comment:
With this removed, some synchronization will need to be done either inside util_ns_add_local_name or in providers that calls that functions otherwise util_ns_add_local_name could be called when the name server is not ready and the local endpoint fails to register with the name server.

Reply:
ns_start initiates the listen before spawning a thread to handle accept.